tools/system_libs.py

```diff
@@ -1388,10 +1388,48 @@ def calculate(temp_files, in_temp, cxx, forced, stdout_=None, stderr_=None):
   if only_forced:
     temp_files = []
 
-  # Add in some hacks for js libraries. If a js lib depends on a symbol provided by a C library, it must be
-  # added to here, because our deps go only one way (each library here is checked, then we check the next
-  # in order - libc++, libcxextra, etc. - and then we run the JS compiler and provide extra symbols from
-  # library*.js files. But we cannot then go back to the C libraries if a new dep was added!
+  # deps_info.json is a mechanism that lets JS code depend on C functions. This
+  # needs special help because of how linking works:
+  #
+  #   1. Receive some input files (.o, .c, etc.) from the user.
+  #   2. Link them with system libraries.
+  #   3. Whatever symbols are still unresolved, look in JS libraries for them.
+  #
+  # This makes C->JS calls work in a natural way: if compiled code depends on
+  # a function foo() that is implemented in a JS library, it will be unresolved
+  # after stage 2, and therefore linked in at stage 3. The problem is the other
+  # direction: if a JS library function decides it needs some function from say
+  # libc, then at stage 3 it is too late to link in more libc code. That's
+  # where deps_info.json comes in.
+  #
+  # Specifically, before stage 2 (linking with system libraries) we look at what
+  # symbols are required by the input files. Imagine that js_func in a JS
+  # library depends on libc_func in libc. Then if deps_info.json tells us
+  #
+  #  "js_func": ["libc_func"]
+  #
+  # then if we see js_func is required (undefined) before stage 2, then we add
+  # a requirement to link in libc_func when linking in system libraries. All
+  # we do with deps_info.json is see if any of the keys are among the
+  # undefined symbols before stage 2, and if they are, add their values to the
+  # things we need to link in.
+  #
+  # This usually works the way you want, but note that it happens *before* stage
+  # 2 and not after it. That is, we look for js_func before linking in system
+  # libraries. If you have a situation where
+  #
+  #   user_code => other_libc_func => js_func => libc_func
+  #
+  # then the deps_info.json entry must contain
+  #
+  #  "other_libc_func": ["libc_func"]
+  #
+  # because that is what we see before stage 2: all we see is that
+  # other_libc_func is going to be linked in, and we don't know yet that it
+  # will end up calling js_func. But the presence of a call to other_libc_func
+  # indicates that we will need libc_func linked in as well, so that is what the
+  # deps_info.json entry should contain.
+  #
```

Comment:
This seems like a really long comment to have inline in the code.  I wonder if there is somewhere else we could document it?    I don't have any great ideas.  Maybe deps_info.txt and refer to that here?.. not ideal I guess

Reply:
Maybe we should write up a document about the entire linking process soon, after the simplifications. And this could go there?